README.rst
aioelasticsearch
================

:info: elasticsearch-py wrapper for asyncio

.. image:: https://img.shields.io/travis/aio-libs/aioelasticsearch.svg
    :target: https://travis-ci.org/aio-libs/aioelasticsearch

.. image:: https://img.shields.io/pypi/v/aioelasticsearch.svg
    :target: https://pypi.python.org/pypi/aioelasticsearch

.. image:: https://codecov.io/gh/aio-libs/aioelasticsearch/branch/master/graph/badge.svg
  :target: https://codecov.io/gh/aio-libs/aioelasticsearch

Installation
------------

For elasticsearch < 7:

.. code-block:: shell

    pip install aioelasticsearch[6]

For elasticsearch >= 7:

.. code-block:: shell

    pip install aioelasticsearch[7]

Usage
-----

.. code-block:: python

    import asyncio

    from aioelasticsearch import Elasticsearch

    async def go():
        es = Elasticsearch()

        print(await es.search())

        await es.close()

    loop = asyncio.get_event_loop()
    loop.run_until_complete(go())
    loop.close()

Features
--------

Asynchronous `scroll <https://www.elastic.co/guide/en/elasticsearch/reference/current/search-request-scroll.html>`_

.. code-block:: python

    import asyncio

    from aioelasticsearch import Elasticsearch
    from aioelasticsearch.helpers import Scan

    async def go():
        async with Elasticsearch() as es:
            async with Scan(
                es,
                index='index',
                doc_type='doc_type',  # omit this for Elasticsearch 7
                query={},
            ) as scan:
                print(scan.total)

                async for doc in scan:
                    print(doc['_source'])

    loop = asyncio.get_event_loop()
    loop.run_until_complete(go())
    loop.close()

Thanks
------

The library was donated by `Ocean S.A. <https://ocean.io/>`_

Thanks to the company for contribution.


CHANGES.rst
Changes
=======

0.7.0 (2019-11-07)
------------------

- Support ``ession_factory_class`` by ``AIOHttpConnection`` (#211)

0.6.0 (2019-11-07)
------------------

- Support elasticsearch 7.x (#194)

0.5.1 (2018-03-12)
------------------

- Don't use deprecated ``verify_ssl`` parameter

0.5.0 (2018-02-14)
------------------

- Fix compatibility with aiohttp 3.0


0.4.0 (2017-12-05)
------------------

- First public release


tests/utils.py
import asyncio


@asyncio.coroutine
def populate(es, index, doc_type, n, body, *, loop):
    coros = []

    yield from es.indices.create(index)

    for i in range(n):
        coros.append(
            es.index(
                index=index,
                doc_type=doc_type,
                id=str(i),
                body=body,
                refresh=True,
            ),
        )

    yield from asyncio.gather(*coros, loop=loop)


tests/__init__.py
import os  # noqa # isort:skip
import logging  # noqa # isort:skip

os.environ['PYTHONUNBUFFERED'] = '1'
os.environ['PYTHONASYNCIODEBUG'] = '1'

logging.basicConfig(level=logging.DEBUG)

import asyncio  # noqa # isort:skip

asyncio.set_event_loop(None)


tests/conftest.py
import asyncio
import gc
import time
import uuid

import elasticsearch
import pytest
from aiohttp.test_utils import unused_port
from docker import from_env as docker_from_env

import aioelasticsearch


@pytest.fixture
def loop(request):
    asyncio.set_event_loop(None)

    loop = asyncio.new_event_loop()

    yield loop

    if not loop._closed:
        loop.call_soon(loop.stop)
        loop.run_forever()
        loop.close()

    gc.collect()
    asyncio.set_event_loop(None)


@pytest.fixture(scope='session')
def session_id():
    '''Unique session identifier, random string.'''
    return str(uuid.uuid4())


@pytest.fixture(scope='session')
def docker():
    client = docker_from_env(version='auto')
    return client


def pytest_addoption(parser):
    parser.addoption("--es_tag", action="append", default=[],
                     help=("Elasticsearch server versions. "
                           "May be used several times. "
                           "6.0.0 by default"))
    parser.addoption("--no-pull", action="store_true", default=False,
                     help="Don't perform docker images pulling")
    parser.addoption("--local-docker", action="store_true", default=False,
                     help="Use 0.0.0.0 as docker host, useful for MacOs X")


def pytest_generate_tests(metafunc):
    if 'es_tag' in metafunc.fixturenames:
        tags = set(metafunc.config.option.es_tag)
        if not tags:
            tags = ['6.0.0']
        else:
            tags = list(tags)
        metafunc.parametrize("es_tag", tags, scope='session')


@pytest.fixture(scope='session')
def es_container(docker, session_id, es_tag, request):
    image = 'docker.elastic.co/elasticsearch/elasticsearch:{}'.format(es_tag)

    if not request.config.option.no_pull:
        docker.images.pull(image)

    es_auth = ('elastic', 'changeme')

    if request.config.option.local_docker:
        es_port_9200 = es_access_port = unused_port()
        es_port_9300 = unused_port()
    else:
        es_port_9200 = es_port_9300 = None
        es_access_port = 9200

    container = docker.containers.run(
        image=image,
        detach=True,
        name='aioelasticsearch-' + session_id,
        ports={
            '9200/tcp': es_port_9200,
            '9300/tcp': es_port_9300,
        },
        environment={
            'http.host': '0.0.0.0',
            'transport.host': '127.0.0.1',
        },
    )

    if request.config.option.local_docker:
        docker_host = '0.0.0.0'
    else:
        inspection = docker.api.inspect_container(container.id)
        docker_host = inspection['NetworkSettings']['IPAddress']

    delay = 0.1
    for _ in range(20):
        es = elasticsearch.Elasticsearch(
            [{
                'host': docker_host,
                'port': es_access_port,
            }],
            http_auth=es_auth,
        )

        try:
            es.transport.perform_request('GET', '/_nodes/_all/http')
        except elasticsearch.TransportError:
            time.sleep(delay)
            delay *= 2
        else:
            break
        finally:
            es.transport.close()
    else:
        pytest.fail("Cannot start elastic server")

    yield {
        'host': docker_host,
        'port': es_access_port,
        'auth': es_auth,
    }

    container.kill(signal=9)
    container.remove(force=True)


@pytest.fixture
def es_clean(es_container):
    def do():
        es = elasticsearch.Elasticsearch(
            hosts=[{
                'host': es_container['host'],
                'port': es_container['port'],
            }],
            http_auth=es_container['auth'],
        )

        try:
            es.transport.perform_request('DELETE', '/_template/*')
            es.transport.perform_request('DELETE', '/_all')
        finally:
            es.transport.close()

    return do


@pytest.fixture
def es_server(es_clean, es_container):
    es_clean()

    return es_container


@pytest.fixture
def es(es_server, auto_close, loop):
    es = aioelasticsearch.Elasticsearch(
        hosts=[{
            'host': es_server['host'],
            'port': es_server['port'],
        }],
        http_auth=es_server['auth'],
        loop=loop,
    )

    return auto_close(es)


@pytest.fixture
def auto_close(loop):
    close_list = []

    def f(arg):
        close_list.append(arg)
        return arg

    yield f

    for arg in close_list:
        loop.run_until_complete(arg.close())


@pytest.mark.tryfirst
def pytest_pycollect_makeitem(collector, name, obj):
    if collector.funcnamefilter(name):
        item = pytest.Function(name, parent=collector)

        if 'run_loop' in item.keywords:
            return list(collector._genfunctions(name, obj))


@pytest.mark.tryfirst
def pytest_pyfunc_call(pyfuncitem):
    if 'run_loop' in pyfuncitem.keywords:
        funcargs = pyfuncitem.funcargs

        loop = funcargs['loop']

        testargs = {
            arg: funcargs[arg]
            for arg in pyfuncitem._fixtureinfo.argnames
        }

        assert asyncio.iscoroutinefunction(pyfuncitem.obj)

        loop.run_until_complete(pyfuncitem.obj(**testargs))

        return True


@pytest.fixture
def populate(es, loop):
    async def do(index, n, body):
        coros = []

        await es.indices.create(index)

        for i in range(n):
            coros.append(
                es.index(
                    index=index,
                    id=str(i),
                    body=body,
                ),
            )

        await asyncio.gather(*coros, loop=loop)
        await es.indices.refresh()
    return do


aioelasticsearch/pool.py
import asyncio
import collections
import logging
import random

from elasticsearch.connection_pool import RoundRobinSelector

from .exceptions import ImproperlyConfigured

logger = logging.getLogger('elasticsearch')


class AIOHttpConnectionPool:

    def __init__(
        self,
        connections,
        dead_timeout=60,
        timeout_cutoff=5,
        selector_class=RoundRobinSelector,
        randomize_hosts=True,
        *,
        loop=None,
        **kwargs
    ):
        self._dead_timeout = dead_timeout
        self.timeout_cutoff = timeout_cutoff
        self.connection_opts = connections
        self.connections = [c for (c, _) in connections]
        self.orig_connections = set(self.connections)
        self.dead = asyncio.PriorityQueue(len(self.connections))
        self.dead_count = collections.Counter()

        self.loop = asyncio.get_event_loop()

        if randomize_hosts:
            random.shuffle(self.connections)

        self.selector = selector_class(dict(connections))

    def dead_timeout(self, dead_count):
        exponent = min(dead_count - 1, self.timeout_cutoff)
        return self._dead_timeout * 2 ** exponent

    def mark_dead(self, connection):
        now = self.loop.time()

        try:
            self.connections.remove(connection)
        except ValueError:
            # connection not alive or marked already, ignore
            return
        else:
            self.dead_count[connection] += 1
            dead_count = self.dead_count[connection]

            timeout = self.dead_timeout(dead_count)

            # it is impossible to raise QueueFull here
            self.dead.put_nowait((now + timeout, connection))

            logger.warning(
                'Connection %r has failed for %i times in a row, '
                'putting on %i second timeout.',
                connection, dead_count, timeout,
            )

    def mark_live(self, connection):
        del self.dead_count[connection]

    def resurrect(self, force=False):
        if self.dead.empty():
            if force:
                # list here is ok, it's a very rare case
                return random.choice(list(self.orig_connections))
            return

        timestamp, connection = self.dead.get_nowait()

        if not force and timestamp > self.loop.time():
            # return it back if not eligible and not forced
            self.dead.put_nowait((timestamp, connection))
            return

        # either we were forced or the connection is elligible to be retried
        self.connections.append(connection)

        logger.info(
            'Resurrecting connection %r (force=%s).',
            connection, force,
        )

        return connection

    def get_connection(self):
        self.resurrect()

        if not self.connections:
            conn = self.resurrect(force=True)
            assert conn is not None
            return conn

        if len(self.connections) > 1:
            return self.selector.select(self.connections)

        return self.connections[0]

    async def close(self, *, skip=frozenset()):
        coros = [
            connection.close() for connection in
            self.orig_connections - skip
        ]

        await asyncio.gather(*coros)


class DummyConnectionPool(AIOHttpConnectionPool):

    def __init__(self, connections, *, loop=None, **kwargs):
        if len(connections) != 1:
            raise ImproperlyConfigured(
                'DummyConnectionPool needs exactly one connection defined.',
            )

        self.loop = asyncio.get_event_loop()

        self.connection_opts = connections
        self.connection = connections[0][0]
        self.connections = [self.connection]
        self.orig_connections = set(self.connections)

    def get_connection(self):
        return self.connection

    async def close(self, *, skip=frozenset()):
        if self.connection in skip:
            return
        await self.connection.close()

    def mark_live(self, connection):
        pass

    def mark_dead(self, connection):
        pass

    def resurrect(self, force=False):
        pass


aioelasticsearch/helpers.py
import logging

from elasticsearch.helpers import ScanError

from aioelasticsearch import NotFoundError

__all__ = ('Scan', 'ScanError')


logger = logging.getLogger('elasticsearch')


class Scan:

    def __init__(
        self,
        es,
        query=None,
        scroll='5m',
        raise_on_error=True,
        preserve_order=False,
        size=1000,
        clear_scroll=True,
        scroll_kwargs=None,
        **kwargs
    ):
        self._es = es

        if not preserve_order:
            query = query.copy() if query else {}
            query['sort'] = '_doc'
        self._query = query
        self._scroll = scroll
        self._raise_on_error = raise_on_error
        self._size = size
        self._clear_scroll = clear_scroll
        self._kwargs = kwargs
        self._scroll_kwargs = scroll_kwargs or {}

        self._scroll_id = None

        self._total = 0

        self._initial = True
        self._done = False
        self._hits = []
        self._hits_idx = 0
        self._successful_shards = 0
        self._total_shards = 0

    async def __aenter__(self):  # noqa
        await self._do_search()
        return self

    async def __aexit__(self, *exc_info):  # noqa
        await self._do_clear_scroll()

    def __aiter__(self):
        if self._initial:
            raise RuntimeError("Scan operations should be done "
                               "inside async context manager")
        return self

    async def __anext__(self):  # noqa
        if self._done:
            raise StopAsyncIteration

        if self._hits_idx >= len(self._hits):
            if self._successful_shards < self._total_shards:
                logger.warning(
                    'Scroll request has only succeeded on %d shards out of %d.',  # noqa
                    self._successful_shards, self._total_shards
                )
                if self._raise_on_error:
                    raise ScanError(
                        self._scroll_id,
                        'Scroll request has only succeeded on {} shards out of {}.'  # noqa
                        .format(self._successful_shards, self._total_shards)
                    )

            await self._do_scroll()
        ret = self._hits[self._hits_idx]
        self._hits_idx += 1
        return ret

    @property
    def scroll_id(self):
        if self._initial:
            raise RuntimeError("Scan operations should be done "
                               "inside async context manager")

        return self._scroll_id

    @property
    def total(self):
        if self._initial:
            raise RuntimeError("Scan operations should be done "
                               "inside async context manager")
        return self._total

    async def _do_search(self):
        self._initial = False

        try:
            resp = await self._es.search(
                body=self._query,
                scroll=self._scroll,
                size=self._size,
                **self._kwargs
            )
        except NotFoundError:
            self._done = True
            return
        else:
            self._total = resp['hits']['total']
            self._update_state(resp)

    async def _do_scroll(self):
        resp = await self._es.scroll(
            scroll_id=self._scroll_id,
            scroll=self._scroll,
            **self._scroll_kwargs,
        )
        self._update_state(resp)

        if self._done:
            raise StopAsyncIteration

    async def _do_clear_scroll(self):
        if self._scroll_id is not None and self._clear_scroll:
            await self._es.clear_scroll(
                body={'scroll_id': [self._scroll_id]},
                ignore=404,
            )

    def _update_state(self, resp):
        self._hits = resp['hits'].get('hits', [])
        self._hits_idx = 0
        self._scroll_id = resp.get('_scroll_id')
        self._successful_shards = resp['_shards']['successful']
        self._total_shards = resp['_shards']['total']
        self._done = not self._hits or self._scroll_id is None


aioelasticsearch/__init__.py
import asyncio

from elasticsearch import Elasticsearch as _Elasticsearch  # noqa # isort:skip
from elasticsearch.connection_pool import (ConnectionSelector, # noqa # isort:skip
                                           RoundRobinSelector)
from elasticsearch.serializer import JSONSerializer  # noqa # isort:skip

from .exceptions import *  # noqa # isort:skip
from .pool import AIOHttpConnectionPool  # noqa # isort:skip
from .transport import AIOHttpTransport  # noqa # isort:skip


__version__ = '0.7.0'


class Elasticsearch(_Elasticsearch):

    def __init__(
        self,
        hosts=None,
        transport_class=AIOHttpTransport,
        *,
        loop=None,
        **kwargs
    ):
        if loop is None:
            loop = asyncio.get_event_loop()

        self.loop = loop

        kwargs['loop'] = self.loop

        super().__init__(hosts, transport_class=transport_class, **kwargs)

    async def close(self):
        await self.transport.close()

    async def __aenter__(self):  # noqa
        return self

    async def __aexit__(self, *exc_info):  # noqa
        await self.close()


aioelasticsearch/transport.py
import asyncio
import logging
from itertools import chain, count

from elasticsearch.serializer import (DEFAULT_SERIALIZERS, Deserializer,
                                      JSONSerializer)
from elasticsearch.transport import Transport, get_host_info

from .connection import AIOHttpConnection
from .exceptions import (ConnectionError, ConnectionTimeout,
                         SerializationError, TransportError)
from .pool import AIOHttpConnectionPool, DummyConnectionPool

logger = logging.getLogger('elasticsearch')


class AIOHttpTransport(Transport):

    def __init__(
        self,
        hosts,
        connection_class=AIOHttpConnection,
        connection_pool_class=AIOHttpConnectionPool,
        host_info_callback=get_host_info,
        serializer=JSONSerializer(),
        serializers=None,
        sniff_on_start=False,
        sniffer_timeout=None,
        sniff_timeout=.1,
        sniff_on_connection_fail=False,
        default_mimetype='application/json',
        max_retries=3,
        retry_on_status=(502, 503, 504, ),
        retry_on_timeout=False,
        send_get_body_as='GET',
        *,
        loop=None,
        **kwargs
    ):
        self.loop = asyncio.get_event_loop()
        self._closed = False

        _serializers = DEFAULT_SERIALIZERS.copy()
        # if a serializer has been specified,
        # use it for deserialization as well
        _serializers[serializer.mimetype] = serializer
        # if custom serializers map has been supplied,
        # override the defaults with it
        if serializers:
            _serializers.update(serializers)
        # create a deserializer with our config
        self.deserializer = Deserializer(_serializers, default_mimetype)

        self.max_retries = max_retries
        self.retry_on_timeout = retry_on_timeout
        self.retry_on_status = retry_on_status
        self.send_get_body_as = send_get_body_as

        # data serializer
        self.serializer = serializer

        # sniffing data
        self.sniffer_timeout = sniffer_timeout
        self.sniff_on_connection_fail = sniff_on_connection_fail
        self.last_sniff = self.loop.time()
        self.sniff_timeout = sniff_timeout

        # callback to construct host dict from data in /_cluster/nodes
        self.host_info_callback = host_info_callback

        # store all strategies...
        self.connection_pool_class = connection_pool_class
        self.connection_class = connection_class
        self._connection_pool_lock = asyncio.Lock()

        # ...save kwargs to be passed to the connections
        self.kwargs = kwargs
        self.hosts = hosts

        # ...and instantiate them
        self.set_connections(hosts)
        # retain the original connection instances for sniffing
        self.seed_connections = set(self.connection_pool.connections)

        self.seed_connection_opts = self.connection_pool.connection_opts

        self.initial_sniff_task = None

        if sniff_on_start:
            def _initial_sniff_reset(fut):
                self.initial_sniff_task = None

            task = self.sniff_hosts(initial=True)

            self.initial_sniff_task = asyncio.ensure_future(task)
            self.initial_sniff_task.add_done_callback(_initial_sniff_reset)

    def set_connections(self, hosts):
        if self._closed:
            raise RuntimeError("Transport is closed")

        def _create_connection(host):
            # if this is not the initial setup look at the existing connection
            # options and identify connections that haven't changed and can be
            # kept around.
            if hasattr(self, 'connection_pool'):
                existing_connections = (self.connection_pool.connection_opts +
                                        self.seed_connection_opts)

                for (connection, old_host) in existing_connections:
                    if old_host == host:
                        return connection

            kwargs = self.kwargs.copy()
            kwargs.update(host)
            kwargs['loop'] = self.loop

            return self.connection_class(**kwargs)

        connections = map(_create_connection, hosts)

        connections = list(zip(connections, hosts))

        if len(connections) == 1:
            self.connection_pool = DummyConnectionPool(
                connections,
                **self.kwargs
            )
        else:
            self.connection_pool = self.connection_pool_class(
                connections,
                loop=self.loop,
                **self.kwargs
            )

    async def _get_sniff_data(self, initial=False):
        previous_sniff = self.last_sniff

        tried = set()

        try:
            # reset last_sniff timestamp
            self.last_sniff = self.loop.time()
            for connection in chain(
                self.connection_pool.connections,
                self.seed_connections,
            ):
                if connection in tried:
                    continue

                tried.add(connection)

                try:
                    # use small timeout for the sniffing request,
                    # should be a fast api call
                    _, headers, node_info = await connection.perform_request(
                        'GET',
                        '/_nodes/_all/http',
                        timeout=self.sniff_timeout if not initial else None,
                    )

                    node_info = self.deserializer.loads(
                        node_info, headers.get('content-type'),
                    )
                    break
                except (ConnectionError, SerializationError):
                    pass
            else:
                raise TransportError('N/A', 'Unable to sniff hosts.')
        except:  # noqa
            # keep the previous value on error
            self.last_sniff = previous_sniff
            raise

        return list(node_info['nodes'].values())

    async def sniff_hosts(self, initial=False):
        if self._closed:
            raise RuntimeError("Transport is closed")
        async with self._connection_pool_lock:
            node_info = await self._get_sniff_data(initial)
            hosts = (self._get_host_info(n) for n in node_info)
            hosts = [host for host in hosts if host is not None]
            # we weren't able to get any nodes, maybe using an incompatible
            # transport_schema or host_info_callback blocked all - raise error.
            if not hosts:
                raise TransportError(
                    'N/A', 'Unable to sniff hosts - no viable hosts found.',
                )

            old_connection_pool = self.connection_pool

            self.set_connections(hosts)

            skip = (self.seed_connections |
                    self.connection_pool.orig_connections)

            await old_connection_pool.close(skip=skip)

    async def close(self):
        if self._closed:
            return
        seeds = self.seed_connections - self.connection_pool.orig_connections

        coros = [connection.close() for connection in seeds]

        if self.initial_sniff_task is not None:
            self.initial_sniff_task.cancel()

            async def _initial_sniff_wrapper():
                try:
                    await self.initial_sniff_task
                except asyncio.CancelledError:
                    return

            coros.append(_initial_sniff_wrapper())

        coros.append(self.connection_pool.close())

        await asyncio.gather(*coros)
        self._closed = True

    async def get_connection(self):
        if self._closed:
            raise RuntimeError("Transport is closed")
        if self.initial_sniff_task is not None:
            await self.initial_sniff_task

        if self.sniffer_timeout:
            if self.loop.time() >= self.last_sniff + self.sniffer_timeout:
                await self.sniff_hosts()

        async with self._connection_pool_lock:
            return self.connection_pool.get_connection()

    async def mark_dead(self, connection):
        if self._closed:
            raise RuntimeError("Transport is closed")
        self.connection_pool.mark_dead(connection)

        if self.sniff_on_connection_fail:
            await self.sniff_hosts()

    async def _perform_request(
        self,
        method, url, params, body,
        ignore=(), timeout=None, headers=None,
    ):
        for attempt in count(1):  # pragma: no branch
            connection = await self.get_connection()

            try:

                status, headers, data = await connection.perform_request(
                    method, url, params, body,
                    ignore=ignore, timeout=timeout, headers=headers,
                )
            except TransportError as e:
                if method == 'HEAD' and e.status_code == 404:
                    return False

                retry = False
                if isinstance(e, ConnectionTimeout):
                    retry = self.retry_on_timeout
                elif isinstance(e, ConnectionError):
                    retry = True
                elif e.status_code in self.retry_on_status:
                    retry = True

                if retry:
                    await self.mark_dead(connection)

                    if attempt == self.max_retries:
                        raise
                else:
                    raise

            else:
                self.connection_pool.mark_live(connection)

                if method == 'HEAD':
                    return 200 <= status < 300

                if data:
                    data = self.deserializer.loads(
                        data, headers.get('content-type'),
                    )

                return data

    async def perform_request(self, method, url, headers=None, params=None, body=None):  # noqa
        if self._closed:
            raise RuntimeError("Transport is closed")
        # yarl fix for https://github.com/elastic/elasticsearch-py/blob/d4efb81b0695f3d9f64784a35891b732823a9c32/elasticsearch/client/utils.py#L29  # noqa
        if params is not None:
            to_replace = {}
            for k, v in params.items():
                if isinstance(v, bytes):
                    to_replace[k] = v.decode('utf-8')
            for k, v in to_replace.items():
                params[k] = v

        if body is not None:
            body = self.serializer.dumps(body)

            # some clients or environments don't support sending GET with body
            if method in ('HEAD', 'GET') and self.send_get_body_as != 'GET':
                # send it as post instead
                if self.send_get_body_as == 'POST':
                    method = 'POST'

                # or as source parameter
                elif self.send_get_body_as == 'source':
                    if params is None:
                        params = {}
                    params['source'] = body
                    params['source_content_type'] = self.serializer.mimetype
                    body = None

        if body is not None:
            try:
                body = body.encode('utf-8', 'surrogatepass')
            except (UnicodeDecodeError, AttributeError):
                # bytes/str - no need to re-encode
                pass

        ignore = ()
        timeout = None
        if params:
            timeout = params.pop('request_timeout', None)
            ignore = params.pop('ignore', ())
            if isinstance(ignore, int):
                ignore = (ignore, )

        return await self._perform_request(
            method, url, params, body,
            ignore=ignore, timeout=timeout, headers=headers,
        )


aioelasticsearch/connection.py
import asyncio

import aiohttp

from .exceptions import ConnectionError, ConnectionTimeout, SSLError  # noqa # isort:skip

from elasticsearch.connection import Connection  # noqa # isort:skip
from yarl import URL  # noqa # isort:skip


def session_factory(**kwargs):
    connector = aiohttp.TCPConnector(
        loop=kwargs.get('loop'),
        limit=kwargs.get('limit', 10),
        use_dns_cache=kwargs.get('use_dns_cache', False),
        ssl=kwargs.get('ssl', False),
    )

    return aiohttp.ClientSession(
        auth=kwargs.get('auth'),
        connector=connector,
    )


class AIOHttpConnection(Connection):

    def __init__(
        self,
        host='localhost',
        port=9200,
        http_auth=None,
        use_ssl=False,
        ssl_context=None,
        verify_certs=False,
        maxsize=10,
        headers=None,
        *,
        loop,
        **kwargs
    ):
        assert not(
            kwargs.get('session') and
            kwargs.get('session_factory')
        ), 'Provide `session` or `session_factory`, not both.'

        super().__init__(host=host, port=port, use_ssl=use_ssl, **kwargs)

        if headers is None:
            headers = {}
        self.headers = headers
        self.headers.setdefault('Content-Type', 'application/json')

        self.loop = loop

        if http_auth is not None:
            if isinstance(http_auth, aiohttp.BasicAuth):
                pass
            elif isinstance(http_auth, str):
                http_auth = aiohttp.BasicAuth(*http_auth.split(':', 1))
            elif isinstance(http_auth, (tuple, list)):
                http_auth = aiohttp.BasicAuth(*http_auth)
            else:
                raise TypeError("Expected str, list, tuple or "
                                "aiohttp.BasicAuth as http_auth parameter,"
                                "got {!r}".format(http_auth))

        self.http_auth = http_auth

        self.verify_certs = verify_certs

        self.base_url = URL.build(scheme='https' if self.use_ssl else 'http',
                                  host=host,
                                  port=port,
                                  path=self.url_prefix)

        self.session = kwargs.get('session')
        self.close_session = False

        if self.session is None:

            self._session_factory = kwargs.get(
                'session_factory',
                session_factory,
            )

            self.session = self._session_factory(
                auth=self.http_auth,
                loop=self.loop,
                ssl=ssl_context if self.verify_certs else False,
                limit=maxsize,
                use_dns_cache=kwargs.get('use_dns_cache', False),
            )

            self.close_session = True

    async def close(self):
        if self.close_session:
            await self.session.close()

    async def perform_request(
        self,
        method,
        url,
        params=None,
        body=None,
        headers=None,
        timeout=None,
        ignore=()
    ):
        url_path = url

        url = (self.base_url / url.lstrip('/')).with_query(params)

        start = self.loop.time()
        try:
            async with self.session.request(
                    method,
                    url,
                    data=body,
                    headers=self._build_headers(headers),
                    timeout=timeout or self.timeout) as response:
                raw_data = await response.text()

                duration = self.loop.time() - start

        except aiohttp.ClientSSLError as exc:
            self.log_request_fail(
                method,
                url,
                url_path,
                body,
                self.loop.time() - start,
                exception=exc,
            )
            raise SSLError('N/A', str(exc), exc)

        except asyncio.TimeoutError as exc:
            self.log_request_fail(
                method,
                url,
                url_path,
                body,
                self.loop.time() - start,
                exception=exc,
            )
            raise ConnectionTimeout('TIMEOUT', str(exc), exc)

        except aiohttp.ClientError as exc:
            self.log_request_fail(
                method,
                url,
                url_path,
                body,
                self.loop.time() - start,
                exception=exc,
            )

            raise ConnectionError('N/A', str(exc), exc)

        # raise errors based on http status codes
        # let the client handle those if needed
        if (
            not (200 <= response.status < 300) and
            response.status not in ignore
        ):
            self.log_request_fail(
                method,
                url,
                url_path,
                body,
                duration,
                response.status,
                raw_data,
            )
            self._raise_error(response.status, raw_data)

        self.log_request_success(
            method,
            url,
            url_path,
            body,
            response.status,
            raw_data,
            duration,
        )

        return response.status, response.headers, raw_data

    def _build_headers(self, headers):
        if headers:
            final_headers = self.headers.copy()
            final_headers.update(headers)
        else:
            final_headers = self.headers
        return final_headers
